protobuf/source_reflection_test.go

```diff
@@ -0,0 +1,74 @@
+package protobuf
+
+import (
+	"net"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+func TestReflection(t *testing.T) {
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	defer ln.Close()
+	require.NoError(t, err)
+	s := grpc.NewServer()
+	reflection.Register(s)
+	go s.Serve(ln)
+
+	source, err := NewDescriptorProviderReflection(ReflectionArgs{
+		Timeout: time.Second,
+		Peers:   []string{ln.Addr().String()},
+	})
+	assert.NoError(t, err)
+	assert.NotNil(t, source)
+
+	result, err := source.FindSymbol("grpc.reflection.v1alpha.ServerReflectionRequest")
+	assert.NoError(t, err)
+	assert.NotNil(t, result)
+
+	result, err = source.FindSymbol("wat")
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "Symbol not found: wat")
+	assert.Nil(t, result)
+}
+
+func TestReflectionMultiplePeers(t *testing.T) {
+	listenRefuser, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "failed to listen on a port")
+	noListen, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "failed to listen on a port")
+	noListen.Close()
+
+	got, err := NewDescriptorProviderReflection(ReflectionArgs{
+		Timeout: time.Second,
+		Peers:   []string{noListen.Addr().String(), listenRefuser.Addr().String()},
+	})
+
+	require.NoError(t, err)
+	require.NotNil(t, got)
+
+	go func() {
+		conn, _ := listenRefuser.Accept()
+		conn.Close()
+	}()
+	_, err = got.FindSymbol("some-symbol")
+	require.Error(t, err)
+}
+
+func TestReflectionClosedPort(t *testing.T) {
+	ln, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err, "failed to listen on a port")
+	ln.Close()
+
+	got, err := NewDescriptorProviderReflection(ReflectionArgs{
+		Timeout: time.Second,
+		Peers:   []string{ln.Addr().String()},
+	})
+
+	assert.Error(t, err)
+	assert.Nil(t, got)
```

Comment:
A variety of these tests assert that the response is `nil` when the error is non-nil. I think we can drop these cases since we only really care whether or not if there was an error. If the error is `non-nil`, the result of the response should be treated as irrelevant and can be discarded.

Reply:
I'm fairly indifferent, but I lean in favour of keeping them because it ensures that we're being idomatic and that no-one will rely on undefined behaviour (of eg a semi initialized struct)

Reply:
I prefer to check if you error that you return nil, just so there's no pointers to half initialised objects lying around.

Reply:
Yea, works for me!